### ISSUE TITLE
Unicode characters as dict keys don't work reliably due to character width ambiguity

### ISSUE THREAD

**Issue by fxkuehl** [numba/numba#6135]
The following code breaks with a KeyError:

```
@njit
def test():
    s = 'a\u1234'
    d = {s[0] : 1}
    print(d['a'])

test()
```

The presence of a wider than 1 byte character in the string "s" triggers the problem. Quote from the documentation:

> String slices also use the same character width as the original string, even if the slice could be represented with a narrower character width. (These details are invisible to the user, of course.)

I suspect that the different character width results in a different hash, which breaks dictionary lookups using such strings as keys. I found this workaround for my application:

```
@njit
def fixuni(c):
    u = c.upper()
    return u if u == c else c.lower()

@njit
def test():
    s = 'a\u1234'
    d = {fixuni(s[0]) : 1}
    print(d['a'])

test()
```

**Comment by esc** [comment]
@fxkuehl thanks for submitting this, your efforts to help to improve Numba are appreciated! RE formatting in GitHub issues, for future reference triple backticks are needed around a code block (with optional language syntax specified), for example: ` ```python <your code> ``` `. GitHub also has a basic guide to its markdown syntax available [here](https://help.github.com/articles/basic-writing-and-formatting-syntax). I have updated the issue for you.

**Comment by esc** [comment]
@fxkuehl thanks for submitting this, I can confirm I am able to reproduce this.

**Comment by seibert** [comment]
I wonder how CPython handles this.  We copied both the string storage logic and the hashing function from Python 3...

**Comment by stuartarchibald** [comment]
Whilst this shows one bug, there's probably two stemming from the same thing. The OP shows a `getitem` with `int`, but a slice would do similar.

------------
Bug 1: string getitem(int) returns a kind the size of the original string.

This is the CPython getitem for an int index:
https://github.com/python/cpython/blob/c3dd7e45cc5d36bbe2295c2840faabb5c75d83e4/Objects/unicodeobject.c#L11985-L12007
note it reads the value based on the kind of the string being index, but returns via `unicode_char`:
https://github.com/python/cpython/blob/c3dd7e45cc5d36bbe2295c2840faabb5c75d83e4/Objects/unicodeobject.c#L2160-L2184
which is based on the minimal kind representation of the character presented.

Numba just reads in the original string size:
https://github.com/numba/numba/blob/b9efc8323c4b5915be55b7a375c4b0b8fdd84fef/numba/cpython/unicode.py#L1639-L1643

------------
Bug 2: string getitem(slice) returns a kind the size of the original string.

For slices, this is the abstract "get a slice from a sequence", note it calls the mapping method `mp_subscript`:
https://github.com/python/cpython/blob/c3dd7e45cc5d36bbe2295c2840faabb5c75d83e4/Objects/abstract.c#L1727-L1748

which for unicode objects is defined here:
https://github.com/python/cpython/blob/c3dd7e45cc5d36bbe2295c2840faabb5c75d83e4/Objects/unicodeobject.c#L14512-L14516

and this is the actual implementation:
https://github.com/python/cpython/blob/c3dd7e45cc5d36bbe2295c2840faabb5c75d83e4/Objects/unicodeobject.c#L14440-L14510

the non-ASCII patch checks the slice for the max char value:
https://github.com/python/cpython/blob/c3dd7e45cc5d36bbe2295c2840faabb5c75d83e4/Objects/unicodeobject.c#L14480-L14491

allocates a new buffer based on this:
https://github.com/python/cpython/blob/c3dd7e45cc5d36bbe2295c2840faabb5c75d83e4/Objects/unicodeobject.c#L14494

and then queries the exact kind and length of the new buffer for doing the write:
https://github.com/python/cpython/blob/c3dd7e45cc5d36bbe2295c2840faabb5c75d83e4/Objects/unicodeobject.c#L14497-L14503

basically, the slice ends up being a string of the minimal kind required to hold the maximum char value.

For a stride 1 slice Numba returns a view, so it'll have the byte width of the original even if the slice doesn't need it. For non-unit stride the allocation will be for a string of the kind of the string being sliced.
https://github.com/numba/numba/blob/b9efc8323c4b5915be55b7a375c4b0b8fdd84fef/numba/cpython/unicode.py#L1646-L1658
